From 2c7c17b95346a3ddafa4e016deb3e32cc2b85418 Mon Sep 17 00:00:00 2001
From: donghuawei <donghuawei@eswincomputing.com>
Date: Fri, 27 Jun 2025 18:35:29 +0800
Subject: [PATCH 591/652] fix: vpu pcie streamid problem

Changelogs:
	vpu pcie card smmu first level ptr point the same
	second ste. and some platform TLP RID is random.
	So do this.

Signed-off-by: donghuawei <donghuawei@eswincomputing.com>
---
 arch/riscv/boot/dts/eswin/smmu_cfg.dtsi      | 256 -------------------
 arch/riscv/configs/vpu7702_evb_defconfig     |   1 +
 arch/riscv/configs/vpu7702_pcie_a2_defconfig |   1 +
 arch/riscv/configs/vpu7702_pcie_defconfig    |   1 +
 drivers/iommu/Kconfig                        |   6 +
 drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c  |  32 ++-
 6 files changed, 39 insertions(+), 258 deletions(-)

diff --git a/arch/riscv/boot/dts/eswin/smmu_cfg.dtsi b/arch/riscv/boot/dts/eswin/smmu_cfg.dtsi
index 0ee4c1ab5eb3..f067df5a1014 100644
--- a/arch/riscv/boot/dts/eswin/smmu_cfg.dtsi
+++ b/arch/riscv/boot/dts/eswin/smmu_cfg.dtsi
@@ -254,259 +254,3 @@
 smmu_cfg_253: smmu_cfg253@253 {};
 smmu_cfg_254: smmu_cfg254@254 {};
 smmu_cfg_255: smmu_cfg255@255 {};
-smmu_cfg_256: smmu_cfg256@256 {};
-smmu_cfg_257: smmu_cfg257@257 {};
-smmu_cfg_258: smmu_cfg258@258 {};
-smmu_cfg_259: smmu_cfg259@259 {};
-smmu_cfg_260: smmu_cfg260@260 {};
-smmu_cfg_261: smmu_cfg261@261 {};
-smmu_cfg_262: smmu_cfg262@262 {};
-smmu_cfg_263: smmu_cfg263@263 {};
-smmu_cfg_264: smmu_cfg264@264 {};
-smmu_cfg_265: smmu_cfg265@265 {};
-smmu_cfg_266: smmu_cfg266@266 {};
-smmu_cfg_267: smmu_cfg267@267 {};
-smmu_cfg_268: smmu_cfg268@268 {};
-smmu_cfg_269: smmu_cfg269@269 {};
-smmu_cfg_270: smmu_cfg270@270 {};
-smmu_cfg_271: smmu_cfg271@271 {};
-smmu_cfg_272: smmu_cfg272@272 {};
-smmu_cfg_273: smmu_cfg273@273 {};
-smmu_cfg_274: smmu_cfg274@274 {};
-smmu_cfg_275: smmu_cfg275@275 {};
-smmu_cfg_276: smmu_cfg276@276 {};
-smmu_cfg_277: smmu_cfg277@277 {};
-smmu_cfg_278: smmu_cfg278@278 {};
-smmu_cfg_279: smmu_cfg279@279 {};
-smmu_cfg_280: smmu_cfg280@280 {};
-smmu_cfg_281: smmu_cfg281@281 {};
-smmu_cfg_282: smmu_cfg282@282 {};
-smmu_cfg_283: smmu_cfg283@283 {};
-smmu_cfg_284: smmu_cfg284@284 {};
-smmu_cfg_285: smmu_cfg285@285 {};
-smmu_cfg_286: smmu_cfg286@286 {};
-smmu_cfg_287: smmu_cfg287@287 {};
-smmu_cfg_288: smmu_cfg288@288 {};
-smmu_cfg_289: smmu_cfg289@289 {};
-smmu_cfg_290: smmu_cfg290@290 {};
-smmu_cfg_291: smmu_cfg291@291 {};
-smmu_cfg_292: smmu_cfg292@292 {};
-smmu_cfg_293: smmu_cfg293@293 {};
-smmu_cfg_294: smmu_cfg294@294 {};
-smmu_cfg_295: smmu_cfg295@295 {};
-smmu_cfg_296: smmu_cfg296@296 {};
-smmu_cfg_297: smmu_cfg297@297 {};
-smmu_cfg_298: smmu_cfg298@298 {};
-smmu_cfg_299: smmu_cfg299@299 {};
-smmu_cfg_300: smmu_cfg300@300 {};
-smmu_cfg_301: smmu_cfg301@301 {};
-smmu_cfg_302: smmu_cfg302@302 {};
-smmu_cfg_303: smmu_cfg303@303 {};
-smmu_cfg_304: smmu_cfg304@304 {};
-smmu_cfg_305: smmu_cfg305@305 {};
-smmu_cfg_306: smmu_cfg306@306 {};
-smmu_cfg_307: smmu_cfg307@307 {};
-smmu_cfg_308: smmu_cfg308@308 {};
-smmu_cfg_309: smmu_cfg309@309 {};
-smmu_cfg_310: smmu_cfg310@310 {};
-smmu_cfg_311: smmu_cfg311@311 {};
-smmu_cfg_312: smmu_cfg312@312 {};
-smmu_cfg_313: smmu_cfg313@313 {};
-smmu_cfg_314: smmu_cfg314@314 {};
-smmu_cfg_315: smmu_cfg315@315 {};
-smmu_cfg_316: smmu_cfg316@316 {};
-smmu_cfg_317: smmu_cfg317@317 {};
-smmu_cfg_318: smmu_cfg318@318 {};
-smmu_cfg_319: smmu_cfg319@319 {};
-smmu_cfg_320: smmu_cfg320@320 {};
-smmu_cfg_321: smmu_cfg321@321 {};
-smmu_cfg_322: smmu_cfg322@322 {};
-smmu_cfg_323: smmu_cfg323@323 {};
-smmu_cfg_324: smmu_cfg324@324 {};
-smmu_cfg_325: smmu_cfg325@325 {};
-smmu_cfg_326: smmu_cfg326@326 {};
-smmu_cfg_327: smmu_cfg327@327 {};
-smmu_cfg_328: smmu_cfg328@328 {};
-smmu_cfg_329: smmu_cfg329@329 {};
-smmu_cfg_330: smmu_cfg330@330 {};
-smmu_cfg_331: smmu_cfg331@331 {};
-smmu_cfg_332: smmu_cfg332@332 {};
-smmu_cfg_333: smmu_cfg333@333 {};
-smmu_cfg_334: smmu_cfg334@334 {};
-smmu_cfg_335: smmu_cfg335@335 {};
-smmu_cfg_336: smmu_cfg336@336 {};
-smmu_cfg_337: smmu_cfg337@337 {};
-smmu_cfg_338: smmu_cfg338@338 {};
-smmu_cfg_339: smmu_cfg339@339 {};
-smmu_cfg_340: smmu_cfg340@340 {};
-smmu_cfg_341: smmu_cfg341@341 {};
-smmu_cfg_342: smmu_cfg342@342 {};
-smmu_cfg_343: smmu_cfg343@343 {};
-smmu_cfg_344: smmu_cfg344@344 {};
-smmu_cfg_345: smmu_cfg345@345 {};
-smmu_cfg_346: smmu_cfg346@346 {};
-smmu_cfg_347: smmu_cfg347@347 {};
-smmu_cfg_348: smmu_cfg348@348 {};
-smmu_cfg_349: smmu_cfg349@349 {};
-smmu_cfg_350: smmu_cfg350@350 {};
-smmu_cfg_351: smmu_cfg351@351 {};
-smmu_cfg_352: smmu_cfg352@352 {};
-smmu_cfg_353: smmu_cfg353@353 {};
-smmu_cfg_354: smmu_cfg354@354 {};
-smmu_cfg_355: smmu_cfg355@355 {};
-smmu_cfg_356: smmu_cfg356@356 {};
-smmu_cfg_357: smmu_cfg357@357 {};
-smmu_cfg_358: smmu_cfg358@358 {};
-smmu_cfg_359: smmu_cfg359@359 {};
-smmu_cfg_360: smmu_cfg360@360 {};
-smmu_cfg_361: smmu_cfg361@361 {};
-smmu_cfg_362: smmu_cfg362@362 {};
-smmu_cfg_363: smmu_cfg363@363 {};
-smmu_cfg_364: smmu_cfg364@364 {};
-smmu_cfg_365: smmu_cfg365@365 {};
-smmu_cfg_366: smmu_cfg366@366 {};
-smmu_cfg_367: smmu_cfg367@367 {};
-smmu_cfg_368: smmu_cfg368@368 {};
-smmu_cfg_369: smmu_cfg369@369 {};
-smmu_cfg_370: smmu_cfg370@370 {};
-smmu_cfg_371: smmu_cfg371@371 {};
-smmu_cfg_372: smmu_cfg372@372 {};
-smmu_cfg_373: smmu_cfg373@373 {};
-smmu_cfg_374: smmu_cfg374@374 {};
-smmu_cfg_375: smmu_cfg375@375 {};
-smmu_cfg_376: smmu_cfg376@376 {};
-smmu_cfg_377: smmu_cfg377@377 {};
-smmu_cfg_378: smmu_cfg378@378 {};
-smmu_cfg_379: smmu_cfg379@379 {};
-smmu_cfg_380: smmu_cfg380@380 {};
-smmu_cfg_381: smmu_cfg381@381 {};
-smmu_cfg_382: smmu_cfg382@382 {};
-smmu_cfg_383: smmu_cfg383@383 {};
-smmu_cfg_384: smmu_cfg384@384 {};
-smmu_cfg_385: smmu_cfg385@385 {};
-smmu_cfg_386: smmu_cfg386@386 {};
-smmu_cfg_387: smmu_cfg387@387 {};
-smmu_cfg_388: smmu_cfg388@388 {};
-smmu_cfg_389: smmu_cfg389@389 {};
-smmu_cfg_390: smmu_cfg390@390 {};
-smmu_cfg_391: smmu_cfg391@391 {};
-smmu_cfg_392: smmu_cfg392@392 {};
-smmu_cfg_393: smmu_cfg393@393 {};
-smmu_cfg_394: smmu_cfg394@394 {};
-smmu_cfg_395: smmu_cfg395@395 {};
-smmu_cfg_396: smmu_cfg396@396 {};
-smmu_cfg_397: smmu_cfg397@397 {};
-smmu_cfg_398: smmu_cfg398@398 {};
-smmu_cfg_399: smmu_cfg399@399 {};
-smmu_cfg_400: smmu_cfg400@400 {};
-smmu_cfg_401: smmu_cfg401@401 {};
-smmu_cfg_402: smmu_cfg402@402 {};
-smmu_cfg_403: smmu_cfg403@403 {};
-smmu_cfg_404: smmu_cfg404@404 {};
-smmu_cfg_405: smmu_cfg405@405 {};
-smmu_cfg_406: smmu_cfg406@406 {};
-smmu_cfg_407: smmu_cfg407@407 {};
-smmu_cfg_408: smmu_cfg408@408 {};
-smmu_cfg_409: smmu_cfg409@409 {};
-smmu_cfg_410: smmu_cfg410@410 {};
-smmu_cfg_411: smmu_cfg411@411 {};
-smmu_cfg_412: smmu_cfg412@412 {};
-smmu_cfg_413: smmu_cfg413@413 {};
-smmu_cfg_414: smmu_cfg414@414 {};
-smmu_cfg_415: smmu_cfg415@415 {};
-smmu_cfg_416: smmu_cfg416@416 {};
-smmu_cfg_417: smmu_cfg417@417 {};
-smmu_cfg_418: smmu_cfg418@418 {};
-smmu_cfg_419: smmu_cfg419@419 {};
-smmu_cfg_420: smmu_cfg420@420 {};
-smmu_cfg_421: smmu_cfg421@421 {};
-smmu_cfg_422: smmu_cfg422@422 {};
-smmu_cfg_423: smmu_cfg423@423 {};
-smmu_cfg_424: smmu_cfg424@424 {};
-smmu_cfg_425: smmu_cfg425@425 {};
-smmu_cfg_426: smmu_cfg426@426 {};
-smmu_cfg_427: smmu_cfg427@427 {};
-smmu_cfg_428: smmu_cfg428@428 {};
-smmu_cfg_429: smmu_cfg429@429 {};
-smmu_cfg_430: smmu_cfg430@430 {};
-smmu_cfg_431: smmu_cfg431@431 {};
-smmu_cfg_432: smmu_cfg432@432 {};
-smmu_cfg_433: smmu_cfg433@433 {};
-smmu_cfg_434: smmu_cfg434@434 {};
-smmu_cfg_435: smmu_cfg435@435 {};
-smmu_cfg_436: smmu_cfg436@436 {};
-smmu_cfg_437: smmu_cfg437@437 {};
-smmu_cfg_438: smmu_cfg438@438 {};
-smmu_cfg_439: smmu_cfg439@439 {};
-smmu_cfg_440: smmu_cfg440@440 {};
-smmu_cfg_441: smmu_cfg441@441 {};
-smmu_cfg_442: smmu_cfg442@442 {};
-smmu_cfg_443: smmu_cfg443@443 {};
-smmu_cfg_444: smmu_cfg444@444 {};
-smmu_cfg_445: smmu_cfg445@445 {};
-smmu_cfg_446: smmu_cfg446@446 {};
-smmu_cfg_447: smmu_cfg447@447 {};
-smmu_cfg_448: smmu_cfg448@448 {};
-smmu_cfg_449: smmu_cfg449@449 {};
-smmu_cfg_450: smmu_cfg450@450 {};
-smmu_cfg_451: smmu_cfg451@451 {};
-smmu_cfg_452: smmu_cfg452@452 {};
-smmu_cfg_453: smmu_cfg453@453 {};
-smmu_cfg_454: smmu_cfg454@454 {};
-smmu_cfg_455: smmu_cfg455@455 {};
-smmu_cfg_456: smmu_cfg456@456 {};
-smmu_cfg_457: smmu_cfg457@457 {};
-smmu_cfg_458: smmu_cfg458@458 {};
-smmu_cfg_459: smmu_cfg459@459 {};
-smmu_cfg_460: smmu_cfg460@460 {};
-smmu_cfg_461: smmu_cfg461@461 {};
-smmu_cfg_462: smmu_cfg462@462 {};
-smmu_cfg_463: smmu_cfg463@463 {};
-smmu_cfg_464: smmu_cfg464@464 {};
-smmu_cfg_465: smmu_cfg465@465 {};
-smmu_cfg_466: smmu_cfg466@466 {};
-smmu_cfg_467: smmu_cfg467@467 {};
-smmu_cfg_468: smmu_cfg468@468 {};
-smmu_cfg_469: smmu_cfg469@469 {};
-smmu_cfg_470: smmu_cfg470@470 {};
-smmu_cfg_471: smmu_cfg471@471 {};
-smmu_cfg_472: smmu_cfg472@472 {};
-smmu_cfg_473: smmu_cfg473@473 {};
-smmu_cfg_474: smmu_cfg474@474 {};
-smmu_cfg_475: smmu_cfg475@475 {};
-smmu_cfg_476: smmu_cfg476@476 {};
-smmu_cfg_477: smmu_cfg477@477 {};
-smmu_cfg_478: smmu_cfg478@478 {};
-smmu_cfg_479: smmu_cfg479@479 {};
-smmu_cfg_480: smmu_cfg480@480 {};
-smmu_cfg_481: smmu_cfg481@481 {};
-smmu_cfg_482: smmu_cfg482@482 {};
-smmu_cfg_483: smmu_cfg483@483 {};
-smmu_cfg_484: smmu_cfg484@484 {};
-smmu_cfg_485: smmu_cfg485@485 {};
-smmu_cfg_486: smmu_cfg486@486 {};
-smmu_cfg_487: smmu_cfg487@487 {};
-smmu_cfg_488: smmu_cfg488@488 {};
-smmu_cfg_489: smmu_cfg489@489 {};
-smmu_cfg_490: smmu_cfg490@490 {};
-smmu_cfg_491: smmu_cfg491@491 {};
-smmu_cfg_492: smmu_cfg492@492 {};
-smmu_cfg_493: smmu_cfg493@493 {};
-smmu_cfg_494: smmu_cfg494@494 {};
-smmu_cfg_495: smmu_cfg495@495 {};
-smmu_cfg_496: smmu_cfg496@496 {};
-smmu_cfg_497: smmu_cfg497@497 {};
-smmu_cfg_498: smmu_cfg498@498 {};
-smmu_cfg_499: smmu_cfg499@499 {};
-smmu_cfg_500: smmu_cfg500@500 {};
-smmu_cfg_501: smmu_cfg501@501 {};
-smmu_cfg_502: smmu_cfg502@502 {};
-smmu_cfg_503: smmu_cfg503@503 {};
-smmu_cfg_504: smmu_cfg504@504 {};
-smmu_cfg_505: smmu_cfg505@505 {};
-smmu_cfg_506: smmu_cfg506@506 {};
-smmu_cfg_507: smmu_cfg507@507 {};
-smmu_cfg_508: smmu_cfg508@508 {};
-smmu_cfg_509: smmu_cfg509@509 {};
-smmu_cfg_510: smmu_cfg510@510 {};
-smmu_cfg_511: smmu_cfg511@511 {};
diff --git a/arch/riscv/configs/vpu7702_evb_defconfig b/arch/riscv/configs/vpu7702_evb_defconfig
index 13c9e8291ed2..bd2d85a0e7cb 100644
--- a/arch/riscv/configs/vpu7702_evb_defconfig
+++ b/arch/riscv/configs/vpu7702_evb_defconfig
@@ -312,6 +312,7 @@ CONFIG_PCIEAER=y
 CONFIG_PCIEASPM_PERFORMANCE=y
 CONFIG_PCIE_PTM=y
 CONFIG_PCIE_ESWIN=y
+CONFIG_ESWIN_PCIE_VPU=y
 CONFIG_DEVTMPFS=y
 CONFIG_DEVTMPFS_MOUNT=y
 CONFIG_MTD=y
diff --git a/arch/riscv/configs/vpu7702_pcie_a2_defconfig b/arch/riscv/configs/vpu7702_pcie_a2_defconfig
index c151e2f411bb..aeee72205a4e 100644
--- a/arch/riscv/configs/vpu7702_pcie_a2_defconfig
+++ b/arch/riscv/configs/vpu7702_pcie_a2_defconfig
@@ -736,6 +736,7 @@ CONFIG_MAILBOX=y
 CONFIG_ESWIN_MBOX=y
 CONFIG_ESWIN_LPCPU=m
 CONFIG_ARM_SMMU_V3=y
+CONFIG_ESWIN_PCIE_VPU=y
 CONFIG_RPMSG_VIRTIO=y
 CONFIG_ARCH_ESWIN_EIC770X_SOC_FAMILY=y
 CONFIG_ESWIN_DSP=m
diff --git a/arch/riscv/configs/vpu7702_pcie_defconfig b/arch/riscv/configs/vpu7702_pcie_defconfig
index c151e2f411bb..eb06fee59706 100644
--- a/arch/riscv/configs/vpu7702_pcie_defconfig
+++ b/arch/riscv/configs/vpu7702_pcie_defconfig
@@ -749,6 +749,7 @@ CONFIG_ESWIN_MMZ_VB=y
 CONFIG_ESWIN_DEV_DMA_BUF=y
 CONFIG_ESWIN_IOMMU_RSV=y
 CONFIG_ESWIN_DMA_MEMCP=y
+CONFIG_ESWIN_PCIE_VPU=y
 CONFIG_PWM=y
 CONFIG_PWM_ESWIN=y
 CONFIG_RESET_ESWIN=y
diff --git a/drivers/iommu/Kconfig b/drivers/iommu/Kconfig
index 7999c71c4612..956c9d54f7a0 100644
--- a/drivers/iommu/Kconfig
+++ b/drivers/iommu/Kconfig
@@ -411,6 +411,12 @@ config ARM_SMMU_V3_SVA
 	  Say Y here if your system supports SVA extensions such as PCIe PASID
 	  and PRI.
 
+config ESWIN_PCIE_VPU
+	bool "Eswin PCIe VPU Card StreamID Macro"
+	depends on ARM_SMMU_V3 || ARCH_ESWIN_EIC770X_SOC_FAMILY
+	help
+	  support for eswin pcie card smmu problem
+
 config S390_IOMMU
 	def_bool y if S390 && PCI
 	depends on S390 && PCI
diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
index 0afb2507c882..39691049512d 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
@@ -1471,6 +1471,28 @@ static void arm_smmu_init_bypass_stes(__le64 *strtab, unsigned int nent, bool fo
 	}
 }
 
+#ifdef CONFIG_ESWIN_PCIE_VPU
+static void eswin_pcie_vpu_smmu_workaround(struct arm_smmu_device *smmu, u32 sid)
+{
+	void *strtab;
+	struct arm_smmu_strtab_cfg *cfg = &smmu->strtab_cfg;
+	struct arm_smmu_strtab_l1_desc *desc = &cfg->l1_desc[sid >> STRTAB_SPLIT];
+	struct arm_smmu_strtab_l1_desc *desc_busx;
+	int i;
+	u32 streamid;
+
+	for (i = 1; i < 256; i++) {
+		streamid = 0xff0000 + (i << 8);
+		strtab = &cfg->strtab[(streamid >> STRTAB_SPLIT) * STRTAB_L1_DESC_DWORDS];
+ 		desc_busx = &cfg->l1_desc[streamid >> STRTAB_SPLIT];
+		desc_busx->span = STRTAB_SPLIT + 1;
+		desc_busx->l2ptr = desc->l2ptr;
+		desc_busx->l2ptr_dma = desc->l2ptr_dma;
+		arm_smmu_write_strtab_l1_desc(strtab, desc_busx);
+	}
+}
+#endif
+
 static int arm_smmu_init_l2_strtab(struct arm_smmu_device *smmu, u32 sid)
 {
 	size_t size;
@@ -1496,6 +1518,12 @@ static int arm_smmu_init_l2_strtab(struct arm_smmu_device *smmu, u32 sid)
 
 	arm_smmu_init_bypass_stes(desc->l2ptr, 1 << STRTAB_SPLIT, false);
 	arm_smmu_write_strtab_l1_desc(strtab, desc);
+
+#ifdef CONFIG_ESWIN_PCIE_VPU
+	if (sid & 0xff0000) {
+		eswin_pcie_vpu_smmu_workaround(smmu, sid);
+	}
+#endif
 	return 0;
 }
 
@@ -1672,9 +1700,9 @@ static irqreturn_t arm_smmu_evtq_thread(int irq, void *dev)
 			if (!ret || !__ratelimit(&rs))
 				continue;
 
-			dev_dbg(smmu->dev, "event 0x%02x received:\n", id);
+			printk("event 0x%02x received:\n", id);
 			for (i = 0; i < ARRAY_SIZE(evt); ++i)
-				dev_dbg(smmu->dev, "\t0x%016llx\n",
+				printk("\t0x%016llx\n",
 					 (unsigned long long)evt[i]);
 
 			cond_resched();
-- 
2.52.0

